src/main/java/com/google/devtools/build/lib/skyframe/RegisteredToolchainsCycleReporter.java

```diff
@@ -0,0 +1,84 @@
+package com.google.devtools.build.lib.skyframe;
+
+import com.google.common.base.Function;
+import com.google.common.base.Predicate;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Iterables;
+import com.google.devtools.build.lib.cmdline.Label;
+import com.google.devtools.build.lib.events.Event;
+import com.google.devtools.build.lib.events.ExtendedEventHandler;
+import com.google.devtools.build.skyframe.CycleInfo;
+import com.google.devtools.build.skyframe.CyclesReporter;
+import com.google.devtools.build.skyframe.SkyKey;
+import java.util.Optional;
+
+public class RegisteredToolchainsCycleReporter implements CyclesReporter.SingleCycleReporter {
+
+  private static final Predicate<SkyKey> IS_REGISTERED_TOOLCHAINS_SKY_KEY =
+      SkyFunctions.isSkyFunction(SkyFunctions.REGISTERED_TOOLCHAINS);
+
+  private static final Predicate<SkyKey> IS_CONFIGURED_TARGET_SKY_KEY =
+      SkyFunctions.isSkyFunction(SkyFunctions.CONFIGURED_TARGET);
+
+  private static final Predicate<SkyKey> IS_TOOLCHAIN_RESOLUTION_SKY_KEY =
+      SkyFunctions.isSkyFunction(SkyFunctions.TOOLCHAIN_RESOLUTION);
+
+  @Override
+  public boolean maybeReportCycle(
+      SkyKey topLevelKey,
+      CycleInfo cycleInfo,
+      boolean alreadyReported,
+      ExtendedEventHandler eventHandler) {
+    ImmutableList<SkyKey> pathToCycle = cycleInfo.getPathToCycle();
+    ImmutableList<SkyKey> cycle = cycleInfo.getCycle();
+    if (pathToCycle.isEmpty()) {
+      return false;
+    } else if (alreadyReported) {
+      return true;
+    } else if (!Iterables.any(cycle, IS_REGISTERED_TOOLCHAINS_SKY_KEY)
+        || !Iterables.any(cycle, IS_CONFIGURED_TARGET_SKY_KEY)
+        || !Iterables.any(cycle, IS_TOOLCHAIN_RESOLUTION_SKY_KEY)) {
+      return false;
+    }
+
+    // Find the ConfiguredTargetKey, this should tell the problem.
```

Comment:
perhaps a nit:

> ... the ConfiguredTargetKey
> ... .findFirst()

note that the order of the cycle is out of the control of this class (depends on blaze-on-skyframe parallelism, the top level keys requests, cached state, etc). do you perhaps mean "Find *a* ConfiguredTargetKey"? or is the cycle expected to contain exactly one ConfiguredTargetKey?

Reply:
There may be cases where there are multiple configured targets, but the correct one to display is the one that is the most direct dependency of RegisteredToolchainFunction.

Reply:
> but the correct one to display is the one that is the most direct dependency of RegisteredToolchainFunction.

so are you going to change the code to do that then?